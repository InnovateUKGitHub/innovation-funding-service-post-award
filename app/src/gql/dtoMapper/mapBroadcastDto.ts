import { BroadcastDto } from "@framework/dtos/BroadcastDto";
import { Clock } from "@framework/util/clock";

/**
 * splits a long string response into an array around new lines
 * with empty lines removed
 */
function parseLongTextArea(unParsedString: string): string[] {
  return unParsedString
    .trim()
    .split("\n")
    .map(x => x.trim())
    .filter(x => x);
}

const clock = new Clock();

type BroadcastNode = Readonly<Partial<{
  Id: string;
  DisplayValue: GQL.Maybe<string>;
  Acc_StartDate__c: GQL.Value<string>;
  Acc_EndDate__c: GQL.Value<string>;
  Acc_Message__c: GQL.Value<string>;
}> | null> | null;

type BroadcastDtoMapping = BroadcastDto;

const mapper: GQL.DtoMapper<BroadcastDtoMapping, BroadcastNode> = {
  id(node) {
    return (node?.Id ?? "") as BroadcastId;
  },
  startDate(node) {
    return !!node?.Acc_StartDate__c?.value
      ? clock.parseRequiredSalesforceDateTime(node.Acc_StartDate__c.value)
      : new Date();
  },
  endDate(node) {
    return !!node?.Acc_EndDate__c?.value
      ? clock.parseRequiredSalesforceDateTime(node.Acc_EndDate__c.value)
      : new Date();
  },
  content(node) {
    return parseLongTextArea(node?.Acc_Message__c?.value ?? "");
  },
  title(node) {
    return node?.DisplayValue ?? "";
  },
};

/**
 * Maps a specified Node from a GQL query to a slice of
 * the BroadcastDto to ensure consistency and compatibility in the application
 */
export function mapToBroadcastDto<T extends BroadcastNode, PickList extends keyof BroadcastDtoMapping>(
  node: T,
  pickList: PickList[],
): Pick<BroadcastDtoMapping, PickList> {
  return pickList.reduce((dto, field) => {
    dto[field] = mapper[field](node);
    return dto;
  }, {} as Pick<BroadcastDtoMapping, PickList>);
}

/**
 * maps from a gql node to an array of broadcast dtos, per picklist
 */
export function mapToBroadcastDtoArray<
  T extends ReadonlyArray<{ node: BroadcastNode } | null> | null,
  PickList extends keyof BroadcastDtoMapping,
>(edges: T, pickList: PickList[]): Pick<BroadcastDtoMapping, PickList>[] {
  return (
    edges?.map(node => {
      return mapToBroadcastDto(node?.node ?? null, pickList);
    }) ?? []
  );
}
