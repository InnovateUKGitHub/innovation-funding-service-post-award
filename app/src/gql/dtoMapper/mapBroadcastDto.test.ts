import { mapToBroadcastDtoArray } from "./mapBroadcastDto";

const data = [
  {
    node: {
      Acc_StartDate__c: {
        value: "2022-02-24T04:00:00.000Z",
      },
      Acc_EndDate__c: {
        value: "2027-03-31T04:00:00.000Z",
      },
      Id: "a2M2600000FmJAcEAN",
      DisplayValue: "Fire!",
      Acc_Message__c: {
        value: "Please leave the building",
      },
    },
  },
  {
    node: {
      Acc_StartDate__c: {
        value: "2022-02-24T04:00:00.000Z",
      },
      Acc_EndDate__c: {
        value: "2027-03-31T04:00:00.000Z",
      },
      Id: "a2M2600000FmJAcEAN",
      DisplayValue: "Nuclear alert",
      Acc_Message__c: {
        value: "Please move to the nearest shelter.",
      },
    },
  },
];

describe("mapToBroadcastDtoArray", () => {
  it("should map the gql data to the dto object", () => {
    expect(mapToBroadcastDtoArray(data, ["id", "title", "content", "startDate", "endDate"])).toMatchSnapshot();
  });
});
