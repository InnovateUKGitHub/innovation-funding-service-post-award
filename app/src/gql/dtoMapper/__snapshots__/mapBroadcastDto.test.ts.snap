// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapToBroadcastDtoArray should map the gql data to the dto object 1`] = `
[
  {
    "content": [
      "Please leave the building",
    ],
    "endDate": 2027-03-31T04:00:00.000Z,
    "id": "a2M2600000FmJAcEAN",
    "startDate": 2022-02-24T04:00:00.000Z,
    "title": "Fire!",
  },
  {
    "content": [
      "Please move to the nearest shelter.",
    ],
    "endDate": 2027-03-31T04:00:00.000Z,
    "id": "a2M2600000FmJAcEAN",
    "startDate": 2022-02-24T04:00:00.000Z,
    "title": "Nuclear alert",
  },
]
`;
